src/utils.js

```diff
@@ -240,13 +251,45 @@ export function getDefaultFormState(_schema, formData, definitions = {}) {
     // No form data? Use schema defaults.
     return defaults;
   }
-  if (isObject(formData)) {
-    // Override schema defaults with form data.
-    return mergeObjects(defaults, formData);
+  if (isObject(formData) || Array.isArray(formData)) {
+    return mergeDefaultsWithFormData(defaults, formData);
   }
   return formData || defaults;
 }
 
+/**
+ * When merging defaults and form data, we want to merge in very specific way:
+ * - objects are deeply merged
+ * - arrays are merged in such way that:
+ *   - when the array is set in form data, only array entries set in form data
+ *     are deeply merged, additional entries from the defaults are ignored
+ *   - when the array is not set in form data, the default is copied over
+ * - scalars are overwritten/set by form data
+ */
+export function mergeDefaultsWithFormData(defaults, formData) {
+  // Recursively merge deeply nested objects.
```

Comment:
Probably don't need this line anymore

Reply:
Good point. 
Do you want me to squash any changes I do or should I do multiple commits?

Reply:
Multiple commits are best -- I'll squash when merging.